// wasm-latex-tools/dist/core/filesystem-manager.js
import { Logger } from '../utils/logger';
export class FileSystemManager {
    constructor(verbose = false) {
        this.logger = new Logger(verbose);
    }
    initialize() {
        if (!window.FS) {
            throw new Error('Emscripten FS not available');
        }
        this.fs = window.FS;
        this.logger.debug('FileSystem initialized');
    }
    writeFile(path, content) {
        try {
            this.ensureDirectory(path);
            this.fs.writeFile(path, content, { encoding: 'utf8' });
            this.logger.debug(`File written: ${path}`);
        }
        catch (error) {
            throw new Error(`Failed to write file ${path}: ${error}`);
        }
    }
    readFile(path) {
        try {
            const result = this.fs.readFile(path, { encoding: 'utf8' });
            return typeof result === 'string' ? result : new TextDecoder().decode(result);
        }
        catch (error) {
            throw new Error(`Failed to read file ${path}: ${error}`);
        }
    }
    fileExists(path) {
        try {
            this.fs.stat(path);
            return true;
        }
        catch {
            return false;
        }
    }
    ensureDirectory(filePath) {
        const dir = filePath.substring(0, filePath.lastIndexOf('/'));
        if (!dir)
            return;
        const parts = dir.split('/').filter(p => p);
        let currentPath = '';
        for (const part of parts) {
            currentPath += '/' + part;
            try {
                this.fs.mkdir(currentPath);
            }
            catch (error) {
                if (error.code && error.code !== 'EEXIST') {
                    throw error;
                }
            }
        }
    }
    deleteFile(path) {
        try {
            if (this.fileExists(path)) {
                this.fs.unlink(path);
                this.logger.debug(`File deleted: ${path}`);
            }
        }
        catch (error) {
            throw new Error(`Failed to delete file ${path}: ${error}`);
        }
    }
}
