// wasm-latex-tools/dist/tools/latexdiff.js
import { BaseTool } from './base-tool';
export class LatexDiff extends BaseTool {
    getScriptPath() {
        return '/latexdiff.pl';
    }
    getDependencyPaths() {
        return [];
    }
    async diff(oldContent, newContent, options) {
        const mergedOptions = {
            input: newContent,
            oldContent,
            ...options
        };
        return this.executeLatexDiff(mergedOptions);
    }
    async executeLatexDiff(options) {
        await this.ensureLoaded();
        const t = Date.now();
        const oldPath = `/tmp/old_${t}.tex`;
        const newPath = `/tmp/new_${t}.tex`;
        const outputPath = `/tmp/diff_${t}.tex`;
        const args = this.buildArguments(oldPath, newPath, outputPath, options);
        const inputs = [
            ...this.preloadedFiles,
            { fn: oldPath, text: options.oldContent },
            { fn: newPath, text: options.input }
        ];
        const outputs = [outputPath];
        return this.runner.runScript(args, inputs, outputs);
    }
    buildArguments(oldPath, newPath, outputPath, options) {
        const diffOptions = options;
        const scriptPath = this.getScriptPath();
        const args = [scriptPath];
        if (diffOptions.type)
            args.push(`--type=${diffOptions.type}`);
        if (diffOptions.subtype)
            args.push(`--subtype=${diffOptions.subtype}`);
        if (diffOptions.floattype)
            args.push(`--floattype=${diffOptions.floattype}`);
        if (diffOptions.encoding)
            args.push(`--encoding=${diffOptions.encoding}`);
        if (diffOptions.excludeSafecmd)
            args.push(`--exclude-safecmd=${diffOptions.excludeSafecmd}`);
        if (diffOptions.appendSafecmd)
            args.push(`--append-safecmd=${diffOptions.appendSafecmd}`);
        if (diffOptions.excludeTextcmd)
            args.push(`--exclude-textcmd=${diffOptions.excludeTextcmd}`);
        if (diffOptions.appendTextcmd)
            args.push(`--append-textcmd=${diffOptions.appendTextcmd}`);
        if (diffOptions.mathMarkup !== undefined)
            args.push(`--math-markup=${diffOptions.mathMarkup}`);
        if (diffOptions.allowSpaces)
            args.push('--allow-spaces');
        if (diffOptions.flatten)
            args.push('--flatten');
        args.push(oldPath, newPath);
        return args;
    }
}
